PIM: Implement AutoRP functionality

nabahr opened the pull request:
This adds most of the AutoRP functionality to PIMD for dynamic RP support using the AutoRP protocol.
RP Discovery is supported and is enabled by default. To disable AutoRP discovery:
```
router pim [vrf NAME]
  no autorp discovery
```
When discovery is enabled, we will listen for the AutoRP Discovery message sent by the AutoRP mapping agent and we will use the discovered RP information.

AutoRP announcement support is also implemented. To configure PIM to advertise a candidate RP via AutoRP:
```
router pim [vrf NAME]
  autorp announce RP-ADDR [GROUP | group-list PREFIX-LIST]
  autorp announce {scope (1-255) | interval (1-65535) | holdtime (0-65535)}
```
The RP advertisement can default to all multicast (224.0.0.0/4), a specific range (225.0.0.0/24), or specified in a prefix list.
The specifics of the announcement messages sent can be configured with the second command. The scope sets the TTL of the packet, the interval sets how often the packet is sent, and the hold time sets the hold time used in the packet, i.e. how long the information in the packet is considered valid (0=disabled).

The only part of AutoRP not yet implemented is the AutoRP mapping agent.

Show information available with:
`show ip pim [vrf NAME] autorp [json]`
This will display state of AutoRP discovery (enabled/disabled) and the configured candidate RP's for AutoRP.
The existing `show ip rp-info` command will display RP's learned via AutoRP.

Resources used for reference during development:
https://github.com/troglobit/pimd/blob/master/doc/pim-autorp-spec01.txt
https://www.cisco.com/c/en/us/td/docs/ios-xml/ios/ipmulti_pim/configuration/imc-pim-xe-3e/imc_autorp.html
https://www.cisco.com/c/en/us/support/docs/ip/multicast/118405-config-rp-00.html

Jafaral commented on tests/topotests/multicast_pim_autorp/test_multicast_pim_autorp.py:
```diff
@@ -0,0 +1,204 @@
+    sleep(10)
+    step("Change AutoRP Announcement packet parameters")
+    rnode.cmd("vtysh -c 'conf t' -c 'router pim' -c 'send-rp-announce scope 8 interval 10 holdtime 60'")
+    step("Verify Announcement sent data")
+    sleep(10)
```
No step here? 

Also, we `sleep()` at lines 189 and 185?

Jafaral commented on tests/topotests/multicast_pim_autorp/test_multicast_pim_autorp.py:
```diff
@@ -0,0 +1,204 @@
+# SPDX-License-Identifier: ISC
+
+#
+# test_multicast_pim_autorp.py
```
can you rename the test to just say `pim_autorp`?

donaldsharp commented on pimd/pim_autorp.c:
```diff
@@ -0,0 +1,617 @@
+
+static void pim_autorp_rp_free(struct pim_autorp_rp *rp)
+{
+	if (rp->hold_timer)
```
EVENT_OFF?

donaldsharp replied:
in any event event_off does not need the test for if the event is scheduled.  It will do the right thing.

nabahr replied:
Sorry, I was just trying to stick to the guidance I found in the libfrr code. I do see that event_cancel does do all the needed checks for NULL though so I will remove that.

```
/*
 * Please consider this macro deprecated, and do not use it in new code.
 */
#define EVENT_OFF(thread)                                                      \
	do {                                                                   \
		if ((thread))                                                  \
			event_cancel(&(thread));                               \
	} while (0)
```

donaldsharp commented on pimd/pim_autorp.c:
```diff
@@ -0,0 +1,617 @@
+	struct pim_instance *pim;
+	struct pim_interface *pim_ifp;
+	pim_addr grp;
+	pim_ifp = ifp->info;
```
We need a empty line here.

donaldsharp commented:
I'm a bit grumpy about the fact that we have basically 3k lines of diff in one big patch.  What can be done to break this up into smaller logical commits?  As it stands it is almost impossible to review given the way it is written.

nabahr committed:
pimd,yang: Implement AutoRP CLI and NB config path

New CLI commands added:
router pim [vrf NAME]
  autorp discovery
  autorp announce RP-ADDR [GROUP | group-list PREFIX-LIST]
  autorp announce {scope (1-255) | interval (1-65535) | holdtime (0-65535)}

autorp discovery
  Enables Auto RP discovery for learning dynamic RP information using the
  AutoRP protocol.

autorp announce RP-ADDR [GROUP | group-list PREFIX-LIST]
  Enable announcements of a candidate RP with the given group range, or
  prefix list of group ranges, to an AutoRP mapping agent.

autorp announce {scope (1-255) | interval (1-65535) | holdtime (0-65535)}
  Configure the parameters of the AutoRP announcement messages.
  The scope sets the packet TTL.
  The interval sets the time between TX of announcements.
  The holdtime sets the hold time in the message, the time the mapping
  agent should wait before invalidating the candidate RP information.

debug pim autorp
  Enable debug logging of the AutoRP protocol

show ip pim [vrf NAME] autorp [json]
  Show details of the AutoRP protocol.
  To view learned RP info, use the existing command 'show ip pim rp-info'

Extend pim yang for new configuration:
  augment /frr-rt:routing/frr-rt:control-plane-protocols/frr-rt:control-plane-protocol/frr-pim:pim/frr-pim:address-family:
    +--rw rp
       +--rw auto-rp
          +--rw discovery-enabled?   boolean
          +--rw announce-scope?      uint8
          +--rw announce-interval?   uint16
          +--rw announce-holdtime?   uint16
          +--rw candidate-rp-list* [rp-address]
             +--rw rp-address           inet:ip-address
             +--rw (group-or-prefix-list)?
                +--:(group)
                |  +--rw group?         frr-route-types:ip-multicast-group-prefix
                +--:(prefix-list)
                   +--rw prefix-list?   plist-ref

Signed-off-by: Nathan Bahr <nbahr@atcorp.com>

nabahr committed:
pimd: Add AutoRP functionality to PIMD

Perform AutoRP discovery and candidate RP announcements using the
AutoRP protocol.
Mapping agent is not yet implemented, but this feature is not
necessary for FRR to support AutoRP as we only need one AutoRP
mapping agent in the network.

Signed-off-by: Nathan Bahr <nbahr@atcorp.com>

nabahr committed:
doc: Add user documentation for AutoRP CLI

Document the CLI commands to configure AutoRP discovery
and candidate RP announcements.

Signed-off-by: Nathan Bahr <nbahr@atcorp.com>

nabahr commented:
> I'm a bit grumpy about the fact that we have basically 3k lines of diff in one big patch. What can be done to break this up into smaller logical commits? As it stands it is almost impossible to review given the way it is written.

The original PR split commits between the implementation of discovery and announcements, but the addition of announcements ended up rewriting parts of the discovery implementation making a review of each commit difficult so I've combined those separate commits and then split that up between the CLI/NB implementation and the actual feature implementation. This splits the size of the change roughly in half. I could try to further decompose the feature implementation but keeping each commit building error free and not adding potential runtime issues to the daemon more difficult.

riw777 reviewed (approved):
looks good, just waiting on comments from other reviewers

donaldsharp commented on pimd/pim_cmd.c:
```diff
@@ -4458,6 +4527,52 @@ DEFPY_ATTR(no_ip_pim_rp_prefix_list,
+}
+
+DEFPY (pim_autorp_announce_rp,
+		 pim_autorp_announce_rp_cmd,
```
can we fix the indentation here?

donaldsharp commented on pimd/pim_cmd.c:
```diff
@@ -4458,6 +4527,52 @@ DEFPY_ATTR(no_ip_pim_rp_prefix_list,
+}
+
+DEFPY (pim_autorp_announce_scope_int,
+		 pim_autorp_announce_scope_int_cmd,
```
fix indentation

donaldsharp commented on pimd/pim_autorp.c:
```diff
@@ -0,0 +1,1146 @@
+	pim_autorp_rp_fini(&(autorp->candidate_rp_list));
+}
+
+static int pim_autorp_join_groups(struct interface *ifp)
```
Why is this a return of a int?  shouldn't it just be a bool? for works/fails?

donaldsharp commented on pimd/pim_autorp.c:
```diff
@@ -0,0 +1,1146 @@
+	return 0;
+}
+
+static int pim_autorp_leave_groups(struct interface *ifp)
```
same why the return of an int -vs- a bool for success failure?

donaldsharp commented on pimd/pim_autorp.c:
```diff
@@ -0,0 +1,1146 @@
+	return 0;
+}
+
+static int pim_autorp_setup(struct pim_autorp *autorp)
```
should just return a bool here imo.

donaldsharp commented on pimd/pim_autorp.c:
```diff
@@ -0,0 +1,1146 @@
+	return 0;
+}
+
+static int pim_autorp_announcement(struct pim_autorp *autorp, uint8_t rpcnt,
```
why have a return?  There is no failure case

donaldsharp replied:
Why was this resolved?  pim_autorp_announcement always returns true?  How can it fail?

nabahr replied:
This is a stub function where the handling of announcement messages will take place when the AutoRP mapping agent is implemented. So right now it doesn't fail because that handling is not implemented but it will have failure cases in the future.

donaldsharp commented on pimd/pim_autorp.c:
```diff
@@ -0,0 +1,1146 @@
+	pim_autorp_rp_free(rp);
+}
+
+static int pim_autorp_add_rp(struct pim_autorp *autorp, pim_addr rpaddr,
```
Isn't thus just success failure?

donaldsharp commented on pimd/pim_autorp.c:
```diff
@@ -0,0 +1,1146 @@
+	return 0;
+}
+
+static int pim_autorp_discovery(struct pim_autorp *autorp, uint8_t rpcnt,
```
Another success/failure function?

donaldsharp commented on pimd/pim_autorp.c:
```diff
@@ -0,0 +1,1146 @@
+	return ret;
+}
+
+static int pim_autorp_msg(struct pim_autorp *autorp, char *buf, size_t buf_size)
```
success failure so bool?

donaldsharp commented on pimd/pim_autorp.c:
```diff
@@ -0,0 +1,1146 @@
+		return -1;
+
+	if (h->type == AUTORP_ANNOUNCEMENT_TYPE &&
+	    pim_autorp_announcement(autorp, h->rpcnt, htons(h->holdtime),
```
The reason I am asking for a bool here is that a -1 as a failure without the function name giving us a clue implies that success is a failure.

donaldsharp commented on pimd/pim_autorp.c:
```diff
@@ -0,0 +1,1146 @@
+	return;
+}
+
+static int pim_autorp_socket_enable(struct pim_autorp *autorp)
```
isn't this just a bool?  The return values are never used

donaldsharp commented on pimd/pim_autorp.c:
```diff
@@ -0,0 +1,1146 @@
+	return 0;
+}
+
+static int pim_autorp_socket_disable(struct pim_autorp *autorp)
```
bool?

donaldsharp commented on pimd/pim_autorp.c:
```diff
@@ -0,0 +1,1146 @@
+
+	brp->addr =
+		rp->addr.s_addr; /* Since this is an in_addr, assume it's already the right byte order */
+	/* TODO What is the right PIM version here? */
```
What is the right version?

donaldsharp commented on pimd/pim_autorp.c:
```diff
@@ -0,0 +1,1146 @@
+	return 0;
+}
+
+int pim_autorp_rm_candidate_rp(struct pim_instance *pim, pim_addr rpaddr)
```
This function never fails?  Why return anything at all?

donaldsharp reviewed (changes requested):
I'm not a fan of negative numbers for success and 0 for failure.  Especially when all that is ever done is did this function succeed fail.  In addition by not testing for negative numbers for success the if statements just looking at them in a vacumn really looks like the function failed.  Please go through and use bool to indicate success/failure and have the functions test appropriately where used.

nabahr committed:
tests: Added tests for new AutoRP functionality

Uses hardcoded sample AutoRP packets injected in to test
message parsing and proper application of AutoRP learned
RP info. Tests mix of AutoRP and static RP's.

Signed-off-by: Nathan Bahr <nbahr@atcorp.com>

nabahr commented:
No argument on the failure checking. I've refactored it to make more sense.
I also fixed the PIM version checking TODO's that got overlooked.
@Jafaral  also asked me to enable AutoRP discovery by default, which matches the behavior of Cisco. This caused a few problems with existing PIM tests so I fixed those in an extra commit.

Jafaral commented on tests/topotests/multicast_pim_sm_topo3/igmp_single_if_group_all_brief.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "totalGroups":5,
+  "totalGroups":"*",
```
drop  the line instead of changing to `*`

nabahr committed:
tests: Update existing tests for changes introduced by AutoRP Discovery

With AutoRP discovery running by default, that adds a new
IGMP group that needs to be accounted for in IGMP output.

For multicast_pim_sm_topo3:
  Ignore the total group number as it is unnecessary for the test.

Signed-off-by: Nathan Bahr <nbahr@atcorp.com>

nabahr committed:
tests: Addition of AutoRP Discovery uncovered broken PIM test

With AutoRP discovery running by default, that adds a new
IGMP group that needs to be accounted for in IGMP output.

For pim.py
  The clear IGMP interfaces function is in a broken state. It was
  already ignoring any errors and returned true always, but with
  the addition of the AutoRP discovery group, you could end up
  with a different group order in the json which would cause a key
  error making the test fail. For now I just added a check to
  avoid the key error.

Signed-off-by: Nathan Bahr <nbahr@atcorp.com>

riw777 reviewed (approved):
changes look good ... waiting on blockers to be cleared and ci to pass

Jafaral commented:
> changes look good ... waiting on blockers to be cleared and ci to pass

passed CI actually.

donaldsharp commented:
Why no testing of sending announcements?  Basically in pim_autorp.c lines 531 - 890 are not even called.  Pim code coverage actually goes down from 54.0% line coverage -> 53.3 % and function coverage is from 63.5% -> 63.0

Large swaths of cli that are added are not exercised at all as well.

nabahr commented:
> Why no testing of sending announcements? Basically in pim_autorp.c lines 531 - 890 are not even called. Pim code coverage actually goes down from 54.0% line coverage -> 53.3 % and function coverage is from 63.5% -> 63.0
> 
> Large swaths of cli that are added are not exercised at all as well.

Manual testing of the announcement messages was done using packet dumps during topotest, but without an AutoRP mapping agent running in the test, the only other way to automate it in topotest would require raw captures to validate the packet data. I couldn't find any examples of other tests that do this or any documentation on how to do this in topotest. 

Completing the AutoRP mapping agent is on the agenda and will be added soon but it is a lower priority at this moment.
Once that implementation is done, these tests will be fully flushed out.

Jafaral commented:
@greptile review this PR. take into account the current state of the code , so don't flag things that are already fixed on the master branch
